Core/SqlStatementModel/ConstantCollectionExpression.cs

```diff
@@ -0,0 +1,109 @@
+﻿// This file is part of the re-linq project (relinq.codeplex.com)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// re-linq is free software; you can redistribute it and/or modify it under
+// the terms of the GNU Lesser General Public License as published by the
+// Free Software Foundation; either version 2.1 of the License,
+// or (at your option) any later version.
+//
+// re-linq is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel
+{
+  /// <summary>
+  /// Expression representing an <see cref="ICollection"/>, an <see cref="ICollection{T}"/> or an <see cref="IReadOnlyCollection{T}"/>.
+  /// </summary>
+  public class ConstantCollectionExpression : Expression
+  {
+    /// <summary>
+    /// The original expression.
+    /// </summary>
+    public ConstantExpression ConstantExpression { get; }
+
+    /// <summary>
+    /// Indicates whether the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s
+    /// <see cref="System.Linq.Expressions.ConstantExpression.Value"/> is a collection without elements.
+    /// </summary>
+    public bool IsEmptyCollection { get; }
+    
+    /// <summary>
+    /// The <see cref="System.Type"/> of the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s <see cref="ConstantExpression.Value"/>.
+    /// </summary>
+    public override Type Type { get; }
+
+    /// <summary>
+    /// Creates a <see cref="ConstantCollectionExpression"/> if the given <paramref name="constantExpression"/> is a collection.
+    /// </summary>
+    /// <exception cref="ArgumentException">
+    ///   If <paramref name="constantExpression"/> is <see langword="null"/>,
+    ///   its <see cref="ConstantExpression.Value"/> is <see langword="null"/>
+    ///   or the <see cref="ConstantExpression.Value"/> is not a collection.
+    /// </exception>
+    public ConstantCollectionExpression (ConstantExpression constantExpression)
```

Comment:
Best as I remember, the ConstantCollectionExpression should get the actual collection objcet in the ctor instead of capturing the original ConstantExpression. We don't wrap the ConstantExpression but rather replace it.

Reply:
The reason we need the original `ConstantExpression` is because of the parameter of `GetOrCreateParameter(ConstantExpression)`, which uses that `ConstantExpression` as key in its `_parameterMap`.

Reply:
Discussed that we check if the ConstantExpression would be de-duplicated by c# compiler. If not, replacing the ConstantExpression with ConstantCollectionExpression will be equivalent. Otherwise, this might be a nice thing but we'd need better names to make it clear that the wrapped ConstantExpression is not part of the ExpressionTree.

Reply:
The compiler creates different expressions with the same value, so unfortunately, there is no re-use possible for these parameters.